### DISCUSSION TITLE
Blog 4

### DISCUSSION THREAD

**Discussion by ktoffel999** [Anastylos/Jump-and-Run#57] [Blog]
Diese Woche haben  wir ein Klassendiagramm erstellt. Zu dem Klassendiagramm mussten wir auch das Gamedesign-Dokument überarbeiten (Link zu unserem SRS: https://github.com/Anastylos/Jump-and-Run/blob/main/Dokumente/SRS_Gamedesign.md)
![Bild_2024-11-05_154020571](https://github.com/user-attachments/assets/4395d4b1-8314-4f9c-81fe-e3606b34588f)

Das Klassendiagramm beschreibt verschiedene Objekte und ihre Rollen im Spiel. Die Item-Klasse stellt Gegenstände dar, die an- oder ausgeschaltet werden können und möglicherweise auf andere Objekte wirken. Die List-Klasse scheint eine Sammlung von Effekten oder Objekten zu sein, die Schaden zufügen, sich bewegen und verschwinden können.
Die Player- und Enemy-Klassen repräsentieren den Spieler und seine Gegner, beide mit Attributen wie Gesundheit, Schaden und Geschwindigkeit. Beide Klassen können sich bewegen, angreifen und sterben, wobei der Spieler zusätzliche Aktionen wie Springen ausführen kann.
Trap und Totem sind statische Objekte mit einem Aktivstatus. Fallen verursachen Schaden, wenn sie aktiviert sind, und Totems könnten bestimmte Effekte auf die Umgebung haben. Die Chest-Klasse repräsentiert Truhen, die vermutlich Belohnungen enthalten oder zerstört werden können.
Die Level-Klasse organisiert das gesamte Level und steuert Ereignisse und Bewegungen über Timer und Zähler, um das Zusammenspiel der Objekte in einem strukturierten Ablauf zu ermöglichen.

Das Dependency Inversion Principle (DIP) wurde in diesem Design angewendet, um die Flexibilität und Erweiterbarkeit des Systems sicherzustellen. Anstatt konkrete Klassen direkt voneinander abhängig zu machen, hängen sie von abstrakten Basisklassen wie Item ab, welche allgemeine Methoden und Eigenschaften definieren (activate(), deactivate()). So können spezifische Klassen wie Trap und Totem unabhängig voneinander implementiert werden, solange sie die allgemeinen Methoden von Item nutzen. Dies ermöglicht es, neue Objekte und Spielmechaniken hinzuzufügen, ohne die bestehenden Klassen zu ändern, da alle Interaktionen durch Abstraktionen statt durch konkrete Implementierungen geregelt werden.

**Comment by OrlandoLegolas** [comment]
Unser Team findet euer Design wirklich beeindruckend! Das Klassendiagramm zeigt eine klare Struktur und macht deutlich, wie vielseitig und modular euer Ansatz ist. Die Anwendung des Dependency Inversion Principle (DIP) sorgt dafür, dass die Architektur flexibel und zukunftssicher bleibt – ein riesiger Vorteil, wenn es um die Erweiterbarkeit des Spiels geht.

Die Abstraktion über die Item-Basisklasse und das klare Rollenverständnis der Klassen wie Player, Enemy, Trap und Chest zeigen, dass ihr großen Wert auf eine gut strukturierte und wartbare Architektur legt. Besonders gefällt uns, dass ihr die Möglichkeit schafft, neue Elemente hinzuzufügen, ohne die Grundstruktur anzutasten. Das wird bei einem komplexen Spielablauf sicher noch wichtig sein. Wir sind gespannt darauf, wie sich das Spiel entwickelt und wie eure Mechaniken und Interaktionen dann im Code und im Gameplay zum Leben erweckt werden!

Viele Grüße 
Das GymTrackerX Team

**Comment by j1tbk** [comment]
Das Klassendiagramm, das hier dargestellt ist, bietet einen strukturierten Überblick über die Beziehungen der einzelnen Klassen in einem Spiel. Die verschiedenen Objekte wie Player, Enemy, Trap, Totem, Chest, und List interagieren auf vielfältige Weise miteinander und mit dem Level. Diese Beziehungen sind durch Assoziationen und Vererbungen visualisiert, um klar zu zeigen, wie die Objekte sich gegenseitig beeinflussen.  Ein gelungenes Beispiel, wie man Komplexität in einem Spielprojekt organisieren kann!

Grüße vom Team HEALTH

**Comment by schmid-di** [comment]
Dieses Klassendiagramm für ein Spielsystem ist ein großartiges Beispiel dafür, wie das Dependency Inversion Principle (DIP) in der Softwarearchitektur angewendet werden kann. Durch die Einführung einer abstrakten Item-Klasse, von der konkrete Objekte wie Trap und Totem erben, wird das System modular und flexibel. Statt dass Klassen direkt voneinander abhängig sind, nutzen sie Abstraktionen, was Erweiterungen und Anpassungen deutlich erleichtert.

Die Hauptklassen wie Player, Enemy, Trap, Totem und Chest erfüllen typische Spielrollen: Der Spieler und seine Gegner können sich bewegen, angreifen und sterben, während statische Objekte wie Fallen und Totems bestimmte Effekte im Spiel haben können. Besonders clever ist die zentrale Level-Klasse, die als Spielmanager fungiert und Timer sowie Zähler für den Ablauf steuert.

Dank der DIP-Anwendung ist das Design leicht erweiterbar, was gerade in der Spieleentwicklung ein enormer Vorteil ist. Neue Mechaniken oder Objekte können hinzugefügt werden, ohne dass bestehende Klassen verändert werden müssen. Das Ergebnis: Ein skalierbares, wartbares System, das auch zukünftigen Anforderungen gerecht wird.

Euer UMOC-Team

/JM /PL
